File: svntrunk/src/untabbed/BlueMatter/probspectrans/src/udf_registryif.cpp
/* Copyright 2001, 2019 IBM Corporation
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the 
 * following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the 
 * following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the 
 * following disclaimer in the documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE 
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 // ***********************************************************************
// File: udf_registryif.cpp
// Author: cwo 
// Date: June 23, 2006
// Namespace: probspectrans 
// Description: 
// ***********************************************************************

#include <BlueMatter/probspectrans/udf_registryif.hpp>

#include <stdlib.h>
#include <assert.h>
#include <iostream.h>

namespace probspectrans
{

//****************************************************************************
// Constructors/Destructors
//****************************************************************************
  UDF_RegistryIF *UDF_RegistryIF::obj_instance = NULL;

  UDF_RegistryIF::UDF_RegistryIF(
          std::string dbname,
          int sys_id,
          RuntimeParams *runtimeparams)
    :  db2conn (db2::DBName::instance(dbname.c_str())->connection()),
       sys_id (sys_id),
       runtimeparams (runtimeparams)
    {
    //empty
    }
  UDF_RegistryIF::~UDF_RegistryIF()
    {
    delete obj_instance;
    obj_instance = NULL;
    }

  UDF_RegistryIF* UDF_RegistryIF::instance(
          std::string dbname,
          int sys_id,
          RuntimeParams *runtimeparams)
    {
    if (obj_instance == NULL)
      {
      obj_instance = new UDF_RegistryIF(dbname, sys_id, runtimeparams);
      assert(obj_instance != NULL);
      }
    return (obj_instance);
    }

//****************************************************************************
//****************************************************************************

  bool UDF_RegistryIF::is_category_for_UDFcode(int code, std::string category)
    {
    db2::StmtHandle stmt1(this->db2conn);
    char sqlcmd1[SQLCMD_BUFSZ];
    snprintf(sqlcmd1, SQLCMD_BUFSZ, "SELECT udf_id FROM Registry.UDF_Category "
            "WHERE udf_id=%d and category_name LIKE '%s' ", code, category.c_str());
    stmt1.prepare(sqlcmd1);

    int ret_code = -1;
    stmt1.bindCol(1, (SQLINTEGER&)ret_code);
    try { stmt1.execute(); stmt1.fetch(); }
    catch(db2::HandleException HE) { std::cerr << HE << " - Caught at Line " << __LINE__ << " in " << __FILE__ << std::endl; }
    
    if (ret_code != -1)
      {
      return true;
      }
    else
      {
      return false;
      }
 
    } //end 


//****************************************************************************
  int UDF_RegistryIF::get_number_of_UDFcodes()
    {
    db2::StmtHandle stmt1(this->db2conn);
    char sqlcmd1[SQLCMD_BUFSZ];
    snprintf(sqlcmd1, SQLCMD_BUFSZ, "SELECT COUNT(*) FROM Registry.UDF_Record");
    stmt1.prepare(sqlcmd1);

    int ret_count = -1;
    stmt1.bindCol(1, (SQLINTEGER&)ret_count);
    try { stmt1.execute(); stmt1.fetch(); }
    catch(db2::HandleException HE) { std::cerr << HE << " - Caught at Line " << __LINE__ << " in " << __FILE__ << std::endl; }

    return ret_count;
    } //end 


//****************************************************************************
  int UDF_RegistryIF::get_UDFcode_by_name(std::string name)
    {
    db2::StmtHandle stmt1(this->db2conn);
    char sqlcmd1[SQLCMD_BUFSZ];
    snprintf(sqlcmd1, SQLCMD_BUFSZ, "SELECT udf_id FROM Registry.UDF_Record "
               " WHERE udf_name LIKE '%s'", name.c_str());
    stmt1.prepare(sqlcmd1);

    int ret_UDFcode = -1;
    stmt1.bindCol(1, (SQLINTEGER&)ret_UDFcode);
    try { stmt1.execute(); stmt1.fetch(); }
    catch(db2::HandleException HE) { std::cerr << HE << " - Caught at Line " << __LINE__ << " in " << __FILE__ << std::endl; }
    
    return ret_UDFcode;
    } //end 


//****************************************************************************
  std::string UDF_RegistryIF::get_UDFname_by_code(int code)
    {
    db2::StmtHandle stmt1(this->db2conn);
    char sqlcmd1[SQLCMD_BUFSZ];
    snprintf(sqlcmd1, SQLCMD_BUFSZ, "SELECT udf_name FROM Registry.UDF_Record "
               " WHERE udf_id=%d ", code);
    stmt1.prepare(sqlcmd1);

    char ret_UDFname[SQLVARCHAR_BUFSZ];
    ret_UDFname[0] = '\n';
    stmt1.bindCol(1, (SQLCHAR*)&ret_UDFname[0], SQLVARCHAR_BUFSZ);
    try { stmt1.execute(); stmt1.fetch(); }
    catch(db2::HandleException HE) { std::cerr << HE << " - Caught at Line " << __LINE__ << " in " << __FILE__ << std::endl; }
 
    std::string ret_str(ret_UDFname);
    return ret_str;
    } //end 


//****************************************************************************
  std::string UDF_RegistryIF::get_structName_by_UDFcode(int code)
    {

    db2::StmtHandle stmt1(this->db2conn);
    char sqlcmd1[SQLCMD_BUFSZ];
    snprintf(sqlcmd1, SQLCMD_BUFSZ, "SELECT udf_name FROM Registry.UDF_Record "
               " WHERE udf_id=%d ", code);
    stmt1.prepare(sqlcmd1);

    char ret_struct_name[SQLVARCHAR_BUFSZ];
    ret_struct_name[0] = '\n';
    stmt1.bindCol(1, (SQLCHAR*)&ret_struct_name[0], SQLVARCHAR_BUFSZ);
    try { stmt1.execute(); stmt1.fetch(); }
    catch(db2::HandleException HE) { std::cerr << HE << " - Caught at Line " << __LINE__ << " in " << __FILE__ << std::endl; }
 
    std::string ret_str(ret_struct_name);
    return ret_str.append("_Params");
    } //end 


//****************************************************************************
  int UDF_RegistryIF::get_Nsites_by_UDFcode(int code) 
    {
    db2::StmtHandle stmt1(this->db2conn);
    char sqlcmd1[SQLCMD_BUFSZ];
    snprintf(sqlcmd1, SQLCMD_BUFSZ, "SELECT site_count FROM Registry.UDF_Record "
               " WHERE udf_id=%d ",code);
    stmt1.prepare(sqlcmd1);

    int ret_count = -1;
    stmt1.bindCol(1, (SQLINTEGER&)ret_count);
    try { stmt1.execute(); stmt1.fetch(); }
    catch(db2::HandleException HE) { std::cerr << HE << " - Caught at Line " << __LINE__ << " in " << __FILE__ << std::endl; }

    return ret_count;
    } //end 


//****************************************************************************
  int UDF_RegistryIF::get_argcount_by_UDFcode(int code)
    {
    db2::StmtHandle stmt1(this->db2conn);
    char sqlcmd1[SQLCMD_BUFSZ];
    snprintf(sqlcmd1, SQLCMD_BUFSZ, "SELECT COUNT( var_name ) FROM "
            " Registry.UDF_Variable WHERE udf_id =%d ", code);
    stmt1.prepare(sqlcmd1);

    int ret_count = -1;
    stmt1.bindCol(1, (SQLINTEGER&)ret_count);
    try { stmt1.execute(); stmt1.fetch(); }
    catch(db2::HandleException HE) { std::cerr << HE << " - Caught at Line " << __LINE__ << " in " << __FILE__ << std::endl; }
    
    return ret_count;
    } //end 


//****************************************************************************
  bool UDF_RegistryIF::reports_energy(std::string name)
    {
    db2::StmtHandle stmt1(this->db2conn);
    char sqlcmd1[SQLCMD_BUFSZ];
    snprintf(sqlcmd1, SQLCMD_BUFSZ, "SELECT count( distinct ruc.udf_id) "
            " FROM registry.udf_record as rur, registry.udf_category as ruc " 
            " WHERE rur.udf_name='%s' and ruc.udf_id=rur.udf_id "
            " and ruc.category_name='ReportsEnergy' ", name.c_str());
    stmt1.prepare(sqlcmd1);

    int ret_count = -1;
    stmt1.bindCol(1, (SQLINTEGER&)ret_count);
    try { stmt1.execute(); stmt1.fetch(); }
    catch(db2::HandleException HE) { std::cerr << HE << " - Caught at Line " << __LINE__ << " in " << __FILE__ << std::endl; }
    
    if (ret_count > 0)
      {
      return true;
      }
    else 
      {
      return false;
      }
    } //end 


//****************************************************************************
  std::string UDF_RegistryIF::get_arg_by_ord(int code, int arg_index)
    {
    db2::StmtHandle stmt1(this->db2conn);
    char sqlcmd1[SQLCMD_BUFSZ];
    snprintf(sqlcmd1, SQLCMD_BUFSZ, "SELECT var_name, type_desc FROM Registry.UDF_Variable " 
            " WHERE udf_id = %d ORDER BY type_size DESC, var_name ASC;", code);
    stmt1.prepare(sqlcmd1);

    char ret_name[SQLVARCHAR_BUFSZ];
    ret_name[0] = '\n';
    stmt1.bindCol(1, (SQLCHAR*)&ret_name[0], SQLVARCHAR_BUFSZ);
    stmt1.execute();

    int count = 0;
    while(stmt1.fetch())
      {
      if (count == arg_index)
        {
        std::string ret_str(ret_name);
        return ret_str;
        }
      count++;
      } 
    
    return ""; 
    } //end 



} //end namespace probspectrans

File: svntrunk/src/BlueMatter/probspecdb2/com/ibm/bluematter/db2probspec/UDF_RegistryIF.java
/* Copyright 2001, 2019 IBM Corporation
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the 
 * following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the 
 * following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the 
 * following disclaimer in the documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE 
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 // ************************************************************************
// File: ProbspecgenDB2.java
// Author: Alex Rayshubskiy
// Date: Sep 12, 2001
// Class: 
// Description: Provides probspecgen the proper access to the db2 UDF_Registry
// Modification:
// Adopted from UDF_RegistryIF.hpp
// ************************************************************************
package com.ibm.bluematter.db2probspec;

import java.sql.*;

public class UDF_RegistryIF {
    
    private Connection con = null;
    
    public static final int CODENOTFOUND = -1;

    public UDF_RegistryIF(Connection c) {
	try {
	    con = c;
	} catch (Exception e) {
	    e.printStackTrace();
	}
    }
    
    // These are all the functionality of UDF_Registry used by probspecgen
//     UDF_RegistryIF::GetUDFCodeByName("LennardJonesForce", ljCode);
//     UDF_RegistryIF::GetStructureStringByUDFCode(ljCode, ljStruct);
//     UDF_RegistryIF::GetStructureNameByUDFCode(ljCode, ljStruct);
//     UDF_RegistryIF::GetNSitesByUDFCode( udfCode, siteCount1 );
//     UDF_RegistryIF::CODENOTFOUND
//     UDF_RegistryIF::GetArgCountByUDFCode(udfCode, argCount);
//     UDF_RegistryIF::GetArgByOrd(udfCode, j, ordName, ordType);
    
    public boolean IsCategoryForCode(int udfCode, String category) {
	try {
	    Statement stmt = con.createStatement();
	    ResultSet rs = stmt.executeQuery("SELECT udf_id FROM Registry.UDF_Category "
					     + "WHERE udf_id=" + udfCode 
					     + " AND category_name LIKE '" + category +"'");
	    
	    if(rs.next()) {
		stmt.close();
		return true;
	    }
	    else {
		stmt.close();
		return false;
	    }	    

	} catch (SQLException e) {
	    e.printStackTrace();	
	}

	return false;
    }
    
    public int GetNumberOfUDFCodes() {
	try {
	    Statement stmt = con.createStatement();
	    ResultSet rs = stmt.executeQuery("SELECT COUNT(*) FROM Registry.UDF_Record");
	    
	    rs.next();
	    int result = rs.getInt(1);
	    stmt.close();
	    return result;
	    
	} catch (SQLException e) {
	    e.printStackTrace();	
	}
	
	return -1;
    }

    public int GetUDFCodeByName(String name) throws Exception {
	try {
	    Statement stmt = con.createStatement();
	    ResultSet rs = stmt.executeQuery("SELECT udf_id FROM Registry.UDF_Record "
					     +"WHERE udf_name LIKE '" + name  + "'");
	    int code;
	    while(rs.next()) {
		code = rs.getInt("udf_id");
		stmt.close();		
		return code;
	    }
	
	    stmt.close();
	    throw new Exception("No udf_id was found for udf_name: " + name);
	    	    
	} catch (SQLException e) {
	    e.printStackTrace();	}

	return -1;
    }

    public String GetUDFNameByCode(int code) throws Exception {
	try {
	    Statement stmt = con.createStatement();
	    ResultSet rs = stmt.executeQuery("SELECT udf_name FROM Registry.UDF_Record "
					     +"WHERE udf_id="+code);
	    String name;
	    while(rs.next()) {
		name = rs.getString(1);
		stmt.close();		
		return name;
	    }
	
	    stmt.close();
	    throw new Exception("No udf_name was found for udf_id: " + code);
	    	    
	} catch (SQLException e) {
	    e.printStackTrace();	}

	return null;
    }

    public String GetStructureStringByUDFCode(int code) throws Exception {
	try {
	    Statement stmt  = con.createStatement();
	    Statement stmt1 = con.createStatement(); 

	    ResultSet rs = stmt.executeQuery("SELECT udf_name FROM Registry.UDF_Record "
					     +"WHERE udf_id =" + code  +";");
	    
	    ResultSet rs1 = stmt1.executeQuery("SELECT var_name, type_desc, type_size FROM Registry.UDF_Variable " 
					      + " WHERE udf_id = " + code + " ORDER BY type_size DESC, var_name ASC;");
	    
	    String structureName = null;
	   
	    while(rs.next()) {
		structureName  = rs.getString("udf_name") + "_Params";
	    }
	    
	    if(structureName == null) {
		stmt.close();
		throw new Exception("No structure was found for udf_id: " + code);
	    }

	    String result = "struct " + structureName + "\n";
	    result += "{\n";
	    while(rs1.next()) {
		result += "\t" + rs1.getString("type_desc") + " " + rs1.getString("var_name") + ";\n";
	    }
	    
	    result += "};\n";
	    
	    stmt.close();
	    return result;
	} catch (SQLException e) {
	    e.printStackTrace();
	}
	return null;
    }

    public String GetStructureNameByUDFCode(int code) throws Exception {
	try {
	    Statement stmt = con.createStatement();

	    ResultSet rs = stmt.executeQuery("SELECT udf_name FROM Registry.UDF_Record "
					     +"WHERE udf_id=" + code  +";");

	    String result = null;
	    while(rs.next()) {
		result = rs.getString("udf_name") + "_Params";
		stmt.close();		
		return result;
	    }
	    
	    stmt.close();
	    throw new Exception("No site_count  was found for udf_id: " + code);
	
	} catch (SQLException e) {
	    e.printStackTrace();
	}

	return null;
    }

    public int GetNSitesByUDFCode(int code) throws Exception {
	try {
	    Statement stmt = con.createStatement();
	    ResultSet rs = stmt.executeQuery("SELECT site_count FROM Registry.UDF_Record "
					     +"WHERE udf_id=" + code  +";");
	    int siteCount;
	    while(rs.next()) {
		siteCount = rs.getInt("site_count");
		stmt.close();		
		return siteCount;
	    }
	    
	    stmt.close();
	    throw new Exception("No site_count  was found for udf_id: " + code);

	} catch (SQLException e) {
	    e.printStackTrace();
	}
	return -1;
    }

    public int GetArgCountByUDFCode(int code) throws Exception {
	try {
	    Statement stmt = con.createStatement();

	    ResultSet rs = stmt.executeQuery("SELECT COUNT( var_name ) FROM Registry.UDF_Variable " 
					     + " WHERE udf_id = " + code + ";");
	    int argCount;
	    while(rs.next()) {
		argCount = rs.getInt(1);
		stmt.close();		
		return argCount;
	    }

	    stmt.close();	    
	    throw new Exception("No arg_count was found for udf_id: " + code);

	} catch (SQLException e) {
	    e.printStackTrace();
	}
	return -1;
    }

    public boolean ReportsEnergy( String name ) throws Exception {
        Statement stmt = con.createStatement();
            
        ResultSet rs = stmt.executeQuery("SELECT count( distinct ruc.udf_id) "
                                         + " FROM registry.udf_record as rur, registry.udf_category as ruc " 
                                         + " WHERE rur.udf_name='"+name+"' and ruc.udf_id=rur.udf_id "
                                         + " and ruc.category_name='ReportsEnergy' ");
        rs.next();
        
        return ( rs.getInt(1) != 0 );
    }
    

    public String GetArgByOrd(int code, int argIndex) throws Exception {
	try {
	    Statement stmt = con.createStatement();
	
	    ResultSet rs = stmt.executeQuery("SELECT var_name, type_desc FROM Registry.UDF_Variable " 
					     + " WHERE udf_id = " + code + " ORDER BY type_size DESC, var_name ASC;");

	    int count = 0;
	    String ordName;
	    while(rs.next()) {
		if (count == argIndex) {
		    ordName = rs.getString("var_name");
		    stmt.close();		
		    return ordName;
		}
		count++;
	    }
	    
	    stmt.close();	    
	    throw new Exception("No ordinal was found for udf_id: " + code + " and index: " + argIndex);
	    
	} catch (SQLException e) {
	    e.printStackTrace();
	}
	
	return null;
    }

    public void close() {
	try {
	    con.close();
	} catch (Exception e) {
	    e.printStackTrace();
	}	
    }
}
